### PR TITLE
add proof-generation crate for masp integration

### PR DESCRIPTION
**Pull request by semaraugusto**
## Summary of changes
- Adds `MaspDelegatedProver` for generating proof in behalf of the user
- Adds `MaspBatchProver` for generating merkle tree batch insertion proofs (works for depositTree, UnspentTree and SpentTree bc the prover doesn't depend on the merkle-tree itself) 
- Adds tests for both provers by providing pre-generated proof inputs

Batch Prover spec
https://www.notion.so/hicommonwealth/Batched-deposit-processing-8a9cbea47acc448cb1076fac86161fd0

Delegated prover spec
https://www.notion.so/hicommonwealth/Delegatable-proof-generation-c8d64aa16957407496e5c706130346d7

### Reference issue to close (if applicable)
- Closes  https://github.com/webb-tools/relayer/issues/441
- Closes https://github.com/webb-tools/relayer/issues/433
- Closes https://github.com/webb-tools/relayer/issues/437
-----
### Code Checklist 

- [ ] Tested
- [ ] Documented

### CODE REVIEW

**Review by drewstone — CHANGES_REQUESTED** [review]
LG

File: crates/circom-proving/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        &proving_key.1,
+        proving_key.1.num_instance_variables,
+        proving_key.1.num_constraints,
+        full_assignment.as_slice(),
+    )?;
+
+    Ok((proof, full_assignment))
+}
+
+/// Verifies a given RLN proof
```
Comment (drewstone) [inline crates/circom-proving/src/lib.rs:117]: Comment need to be updated.

**Comment by shekohex** [comment]
It would be good to have a meeting or a call about this PR and the overall task so I could understand better and review the PR with the correct assumptions.

**Review by shekohex — COMMENTED** [review]
Looks good to me! 
It needs a little bit of clean ups but we can do that later on new PRs.

**Pull request merged**